### PR TITLE
BLD: basic sinewave

### PR DESCRIPTION
**Pull request by eigenfoo**
Supercedes #92

### CODE REVIEW

File: sandbox/mooncake.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    ):
+        super(Mooncake, self).__init__()
+        self.num_channels = num_channels
+        self.num_blocks = num_blocks
+
+        self.conv1 = CausalConv1d(
+            in_channels, num_channels, kernel_size=kernel_size, dilation=1)
+        self.conv2 = nn.Conv1d(num_channels, 2 * num_channels, kernel_size=1)
+
+        self.blocks = [
```
Comment (ccurro) [inline sandbox/mooncake.py]: wrap with `nn.ModuleList`

File: sandbox/mooncake.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    Returns
+    -------
+    sinewave : [batch_size, 100]
+    """
+    sinewaves = []
+    for i in range(batch_size):
+        n = 100  # FIXME np.random.randint(50, 150)
+        f = 1
+        phi = 0
+        x = np.linspace(0, 2 / f, n, dtype=np.float32)
```
Comment (ccurro) [inline sandbox/mooncake.py]: move 2 to `num_periods`

File: sandbox/mooncake.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        sinewave = np.sin(2 * np.pi * f * (x + phi))
+        pad_length = max_num_samples - len(sinewave)
+        sinewave = np.pad(
+            sinewave, (0, pad_length), "constant", constant_values=0)
+        sinewave += 0.1 * np.random.randn(max_num_samples)
```
Comment (ccurro) [inline sandbox/mooncake.py]: don't hardcode variance

File: sandbox/mooncake.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    pi, mu, sigma : [batch_size, num_components (+1 for pi), max_num_samples]
+        Mixture weights, means and standard deviations of each Gaussian
+        component, respectively.
+
+    Returns
+    -------
+    [batch_size, max_num_samples - 1]
+    """
+    # Index appropriately to compute nlogp of predicted next value.
+    vals = x.unsqueeze(1)[..., 1:]
```
Comment (ccurro) [inline sandbox/mooncake.py]: personally i'd unsqueeze in the data gen func rather than in multiple places in the forward pass/loss calc

Comment (ccurro) [reply sandbox/mooncake.py]: you can see how i did it in ar-exp

File: sandbox/mooncake.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+class Mooncake(nn.Module):
+    # TODO tweak parameters
+    def __init__(
+            self,
+            in_channels=1,
+            num_channels=4,
+            num_blocks=8,
+            kernel_size=2,
+            dilations=[1, 2, 4, 8, 16, 32, 64, 128],
```
Comment (ccurro) [inline sandbox/mooncake.py]: num_blocks needs to match length of dilations. either error check or make it so it happens automatically

File: sandbox/mooncake.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            Mixture weights, means and standard deviations of each Gaussian
+            component, respectively.
+        """
+        x = x.unsqueeze(1)  # Sinewave only has 1 channel
+        x = self.conv1(x)
+
+        residual = x
+        for i in range(self.num_blocks):
+            residual = self.blocks[i](residual)
+            x += residual
```
Comment (ccurro) [inline sandbox/mooncake.py]: dedup some variable names

File: sandbox/mooncake.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+    mooncake = Mooncake().to(DEVICE)
+    optimizer = optim.Adam(mooncake.parameters(), lr=0.001)
+
+    # Train
+    mooncake.train()
+
+    for i in range(NUM_TRAIN_ITERATIONS):
+        # Generate one sinewave and make a batch out of it.
+        sinewaves = generate_sinewaves(1, max_num_samples=max_num_samples)
```
Comment (ccurro) [inline sandbox/mooncake.py]: `batch_size` not 1

File: sandbox/mooncake.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        pi, mu, sigma = mooncake(sinewaves)
+
+        # Update
+        optimizer.zero_grad()
+        nlogp = negative_log_prob(sinewaves, pi, mu, sigma).sum(dim=1).mean()
```
Comment (ccurro) [inline sandbox/mooncake.py]: just change .sum(dim=1).mean to .mean. it plays better with adam even if it's not technically "correct"

File: sandbox/mooncake.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        inferred = torch.zeros_like(ground_truth)
+        inferred[..., :SEED_LENGTH] = ground_truth[..., :SEED_LENGTH]
+
+        for i in range(SEED_LENGTH, inferred.shape[1]):
+            pi, mu, sigma = mooncake(inferred[..., :i])
+
+            pi = pi[:, 1:, -1]
+            mu = mu[..., -1]
+            sigma = sigma[..., -1]
+            inferred[..., i] = (pi * mu).sum(dim=1)
```
Comment (ccurro) [inline sandbox/mooncake.py]: fine for now. but keep in mind we might not always be inferring the mean of the gmm

File: sandbox/mooncake.py (hunk 8 above)
Comment (eigenfoo) [reply sandbox/mooncake.py]: @ccurro can you explain why it plays better with adam?

Comment (ccurro) [reply sandbox/mooncake.py]: Actually adam is scale invariant. So disregard the later part of my comment.

It'll be nice to report the mean here since we'll be reporting average likelihood score per sample regardless of how many samples an example takes up. We don't want examples with different lengths to have different likelihood scores just b/c they're different lengths - that makes it harder to compare for us.

File: sandbox/mooncake.py [hunk 10]
```diff
@@ -180,20 +191,20 @@ def forward(self, x):
...
+        residual = h1
         for i in range(self.num_blocks):
             residual = self.blocks[i](residual)
-            x += residual
+            y += residual
 
-        z = F.relu(x)
-        z = self.conv2(z)
+        z = F.relu(y)
+        h2 = self.conv2(z)
```
Comment (ccurro) [inline sandbox/mooncake.py]: we go from h1 -> y -> z -> h2 

:face_with_head_bandage:

Comment (eigenfoo) [reply sandbox/mooncake.py]: I thought you said outputs of convs were `h`s??
